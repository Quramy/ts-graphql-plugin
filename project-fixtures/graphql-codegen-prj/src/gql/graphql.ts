/* eslint-disable */
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
};

export type Post = {
  __typename?: 'Post';
  author: User;
  body: Scalars['String']['output'];
  createdAt: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  title: Scalars['String']['output'];
  updatedAt: Scalars['String']['output'];
};

export type Query = {
  __typename?: 'Query';
  popularPosts: Array<Post>;
};

export type User = {
  __typename?: 'User';
  avatarURL: Scalars['String']['output'];
  createdAt: Scalars['String']['output'];
  id: Scalars['ID']['output'];
  name: Scalars['String']['output'];
  updatedAt: Scalars['String']['output'];
};

export type PopularPosts_QueryQueryVariables = Exact<{ [key: string]: never; }>;


export type PopularPosts_QueryQuery = { __typename?: 'Query', popularPosts: Array<(
    { __typename?: 'Post', id: string }
    & { ' $fragmentRefs'?: { 'PostSummary_PostFragment': PostSummary_PostFragment } }
  )> };

export type PostSummary_PostFragment = { __typename?: 'Post', id: string, title: string, author: (
    { __typename?: 'User', name: string }
    & { ' $fragmentRefs'?: { 'UserAvatar_UserFragment': UserAvatar_UserFragment } }
  ) } & { ' $fragmentName'?: 'PostSummary_PostFragment' };

export type UserAvatar_UserFragment = { __typename?: 'User', name: string, avatarURL: string } & { ' $fragmentName'?: 'UserAvatar_UserFragment' };

export const UserAvatar_UserFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}}]} as unknown as DocumentNode<UserAvatar_UserFragment, unknown>;
export const PostSummary_PostFragmentDoc = {"kind":"Document","definitions":[{"kind":"FragmentDefinition","name":{"kind":"Name","value":"PostSummary_Post"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Post"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"author"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"UserAvatar_User"}}]}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}}]} as unknown as DocumentNode<PostSummary_PostFragment, unknown>;
export const PopularPosts_QueryDocument = {"kind":"Document","definitions":[{"kind":"OperationDefinition","operation":"query","name":{"kind":"Name","value":"PopularPosts_Query"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"popularPosts"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"PostSummary_Post"}}]}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"UserAvatar_User"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"User"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"Field","name":{"kind":"Name","value":"avatarURL"}}]}},{"kind":"FragmentDefinition","name":{"kind":"Name","value":"PostSummary_Post"},"typeCondition":{"kind":"NamedType","name":{"kind":"Name","value":"Post"}},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"id"}},{"kind":"Field","name":{"kind":"Name","value":"title"}},{"kind":"Field","name":{"kind":"Name","value":"author"},"selectionSet":{"kind":"SelectionSet","selections":[{"kind":"Field","name":{"kind":"Name","value":"name"}},{"kind":"FragmentSpread","name":{"kind":"Name","value":"UserAvatar_User"}}]}}]}}]} as unknown as DocumentNode<PopularPosts_QueryQuery, PopularPosts_QueryQueryVariables>;