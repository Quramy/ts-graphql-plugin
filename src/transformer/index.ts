export { TransformerHost, GetTransformerOptions } from './transformer-host';
