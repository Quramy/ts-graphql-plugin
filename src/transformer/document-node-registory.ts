import { DocumentNode } from 'graphql';
import { ExtractResult } from '../analyzer/extractor';

export class DocumentNodeRegistory {
  protected readonly _map = new Map<string, Map<number, DocumentNode>>();

  constructor() {}

  getFiles() {
    return [...this._map.keys()];
  }

  getDocumentNode(templateNode: ts.TemplateExpression | ts.NoSubstitutionTemplateLiteral) {
    const positionMap = this._map.get(templateNode.getSourceFile().fileName);
    if (!positionMap) return;
    return positionMap.get(templateNode.getStart());
  }

  deleteDocumentNodesByFileName(fileName: string) {
    this._map.delete(fileName);
    return this;
  }

  update(extractedResults: ExtractResult[]) {
    extractedResults.forEach(result => {
      if (!result.documentNode) return;
      let positionMap = this._map.get(result.fileName);
      if (!positionMap) {
        positionMap = new Map<number, DocumentNode>();
        this._map.set(result.fileName, positionMap);
      }
      positionMap.set(result.templateNode.getStart(), result.documentNode);
    });
  }
}
