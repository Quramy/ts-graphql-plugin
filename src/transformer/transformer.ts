import ts from 'typescript';
import { DocumentNode, print } from 'graphql';
import { hasTagged, removeAliasFromImportDeclaration } from '../ts-ast-util';

export type DocumentTransformer = (documentNode: DocumentNode) => DocumentNode;

export type TransformOptions = {
  tag?: string;
  documentTransformers: DocumentTransformer[];
  removeFragmentDefinitions: boolean;
  target: 'text' | 'object';
  getDocumentNode: (node: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression) => DocumentNode | undefined;
  getEnabled: () => boolean;
};

function toObjectNode(field: any): ts.Expression {
  if (field === null) {
    return ts.createNull();
  } else if (typeof field === 'boolean') {
    return field ? ts.createTrue() : ts.createFalse();
  } else if (typeof field === 'number') {
    return ts.createNumericLiteral(field + '');
  } else if (typeof field === 'string') {
    return ts.createStringLiteral(field);
  } else if (Array.isArray(field)) {
    return ts.createArrayLiteral(field.map(item => toObjectNode(item)));
  }
  return ts.createObjectLiteral(
    Object.entries(field)
      .filter(([k, v]) => k !== 'loc' && v !== undefined)
      .map(([k, v]) => ts.createPropertyAssignment(ts.createIdentifier(k), toObjectNode(v))),
    true,
  );
}

export function getTransformer({
  tag,
  target,
  getDocumentNode,
  removeFragmentDefinitions,
  documentTransformers,
  getEnabled,
}: TransformOptions) {
  return (ctx: ts.TransformationContext) => {
    const visit = (node: ts.Node): ts.Node | undefined => {
      if (!getEnabled()) return node;
      let templateNode: ts.NoSubstitutionTemplateLiteral | ts.TemplateExpression | undefined = undefined;

      if (tag && ts.isImportDeclaration(node)) {
        return removeAliasFromImportDeclaration(node, tag);
      }

      if (ts.isTaggedTemplateExpression(node) && (!tag || hasTagged(node, tag))) {
        templateNode = node.template;
      } else if (!tag && ts.isNoSubstitutionTemplateLiteral(node)) {
        templateNode = node;
      } else if (!tag && ts.isTemplateExpression(node)) {
        templateNode = node;
      }

      if (!templateNode) return ts.visitEachChild(node, visit, ctx);

      const originalDocumentNode = getDocumentNode(templateNode);
      if (!originalDocumentNode) return ts.visitEachChild(node, visit, ctx);
      const documentNode = documentTransformers.reduce((doc, dt) => dt(doc), originalDocumentNode);
      if (!documentNode) return ts.visitEachChild(node, visit, ctx);
      const toBeRemoved =
        removeFragmentDefinitions && documentNode.definitions.every(def => def.kind === 'FragmentDefinition');
      if (target === 'text') {
        if (toBeRemoved) return ts.createStringLiteral('');
        return ts.createStringLiteral(print(documentNode));
      }
      if (toBeRemoved) return ts.createNumericLiteral('0');
      return toObjectNode(documentNode);
    };
    return (sourceFile: ts.SourceFile) => ts.visitEachChild(sourceFile, visit, ctx);
  };
}
