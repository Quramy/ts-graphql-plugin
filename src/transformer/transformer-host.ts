import { DocumentNode } from 'graphql';
import { Analyzer, AnalyzerFactory, ExtractResult } from '../analyzer';
import { getTransformer, DocumentTransformer } from './transformer';

class DocumentNodeRegistory {
  protected readonly _map = new Map<string, Map<number, DocumentNode>>();

  constructor() {}

  getFiles() {
    return [...this._map.keys()];
  }

  getDocumentNode(templateNode: ts.TemplateExpression | ts.NoSubstitutionTemplateLiteral) {
    const positionMap = this._map.get(templateNode.getSourceFile().fileName);
    if (!positionMap) return;
    return positionMap.get(templateNode.getStart());
  }

  update(extractedResults: ExtractResult[]) {
    extractedResults.forEach(result => {
      if (!result.documentNode) return;
      let positionMap = this._map.get(result.fileName);
      if (!positionMap) {
        positionMap = new Map<number, DocumentNode>();
        this._map.set(result.fileName, positionMap);
      }
      positionMap.set(result.templateNode.getStart(), result.documentNode);
    });
  }
}

export type CreateTransformServerOptions = {
  projectPath: string;
};

export type GetTransformerOptions = {
  target?: 'object' | 'text';
  removeFragmentDefinitions?: boolean;
  documentTransformers?: string[];
  getEnabled?: () => boolean;
};

export class TransformerHost {
  private readonly _analyzer: Analyzer;
  private readonly _scriptHost: { reloadFile: (fileName: string) => void };
  private readonly _documentNodeRegistory = new DocumentNodeRegistory();

  constructor({ projectPath }: CreateTransformServerOptions) {
    const { analyzer, scriptHost } = new AnalyzerFactory().createAnalyzerAndScriptHostFromProjectPath(projectPath);
    this._analyzer = analyzer;
    this._scriptHost = { reloadFile: scriptHost.loadFromFileSystem.bind(scriptHost) };
  }

  loadProject() {
    const [, results] = this._analyzer.extract();
    this._documentNodeRegistory.update(results);
  }

  updateFiles(fileNameList: string[]) {
    fileNameList.forEach(fileName => this._scriptHost.reloadFile(fileName));

    // Note:
    // We need re-extract from not only changed .ts files but other files already opened
    // because the operations/fragments in not changed files can be affected by the change.
    //
    // For example:
    //    changed-file.ts     : export `fragment X on Query { fieldA }`
    //    other-opened-file.ts: declare `query { ...X }` importing fragment from changed-file.ts
    //
    // In the above case, the transformed output of other-opened-file.ts should have GraphQL docuemnt corresponding to `fragment X on Query { fieldA } query { ...X }`
    const [, results] = this._analyzer.extract([
      ...new Set([...fileNameList, ...this._documentNodeRegistory.getFiles()]),
    ]);
    this._documentNodeRegistory.update(results);
  }

  getTransformer({
    target = 'object',
    removeFragmentDefinitions = true,
    documentTransformers = [],
    getEnabled = () => true,
  }: GetTransformerOptions = {}) {
    const { tag } = this._analyzer.getPluginConfig();
    const documentTransformerModules = documentTransformers.map(transformerName => {
      const mod = require(transformerName);
      if (typeof mod !== 'function') {
        throw new Error('Document transofmer should be function');
      }
      return mod as DocumentTransformer;
    });
    return getTransformer({
      getEnabled,
      tag,
      target,
      removeFragmentDefinitions,
      getDocumentNode: node => this._documentNodeRegistory.getDocumentNode(node),
      documentTransformers: documentTransformerModules,
    });
  }
}
