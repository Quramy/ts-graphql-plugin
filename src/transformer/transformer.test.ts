import ts from 'typescript';
import { DocumentNode, parse, visit } from 'graphql';

import { getTransformer } from './transformer';

function transformAndPrint({
  tag,
  target,
  docContent,
  tsContent,
  removeFragmentDefinitions = true,
  documentTransformers = [],
  enabled = true,
}: {
  tag?: string;
  target: 'text' | 'object';
  docContent: string;
  tsContent: string;
  removeFragmentDefinitions?: boolean;
  documentTransformers?: ((doc: DocumentNode) => DocumentNode)[];
  enabled?: boolean;
}) {
  const getDocumentNode = () => parse(docContent);
  const source = ts.createSourceFile('main.ts', tsContent, ts.ScriptTarget.Latest, true);
  const transformer = getTransformer({
    tag,
    target,
    getDocumentNode,
    removeFragmentDefinitions,
    documentTransformers,
    getEnabled: () => enabled,
  });
  const { transformed } = ts.transform(source, [transformer]);
  return ts.createPrinter({ newLine: ts.NewLineKind.LineFeed }).printFile(transformed[0]);
}

describe('transformer', () => {
  describe('GraphQL document transformation', () => {
    it('should transform TaggedTemplateExpression', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = hoge\`abc\`;
        `,
          docContent: `
          query {
            hello
          }
        `,
          target: 'object',
        }),
      ).toMatchSnapshot();
    });

    it('should transform NoSubstitutionTemplateLiteral', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = \`abc\`;
        `,
          docContent: `
          query {
            hello
          }
        `,
          target: 'object',
        }),
      ).toMatchSnapshot();
    });

    it('should transform TemplateExpression', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = \`abc\${def}\`;
        `,
          docContent: `
          query {
            hello
          }
        `,
          target: 'object',
        }),
      ).toMatchSnapshot();
    });

    it('should ignore TaggedTemplateExpression when the node does not match tag name', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = hoge\`abc\`;
        `,
          tag: 'foo',
          docContent: `
          query {
            hello
          }
        `,
          target: 'object',
        }),
      ).toMatchSnapshot();
    });

    it('should transform TaggedTemplateExpression when the node matches tag name', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = hoge\`abc\`;
        `,
          tag: 'hoge',
          docContent: `
          query {
            hello
          }
        `,
          target: 'object',
        }),
      ).toMatchSnapshot();
    });

    it('should ignore NoSubstitutionTemplateLiteral when tag is set', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = \`abc\`;
        `,
          tag: 'foo',
          docContent: `
          query {
            hello
          }
        `,
          target: 'object',
        }),
      ).toMatchSnapshot();
    });

    it('should ignore TemplateExpression when tag is set', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = \`abc\${def}\`;
        `,
          tag: 'foo',
          docContent: `
          query {
            hello
          }
        `,
          target: 'object',
        }),
      ).toMatchSnapshot();
    });

    it('should transform to 0 literal when removeFragmentDefinitions: true and document has only fragments', () => {
      expect(
        transformAndPrint({
          tsContent: `
            const fragment = hoge\`abc\`;
          `,
          docContent: `
            fragment X on Query {
              helo
            }
          `,
          target: 'object',
          removeFragmentDefinitions: true,
        }),
      ).toMatchSnapshot();
    });

    it('should transform to string literal when target is text', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = hoge\`abc\`;
        `,
          docContent: `
          query MyQuery {
            hello
          }
        `,
          target: 'text',
        }),
      ).toMatchSnapshot();
    });

    it('should transform to empty string when removeFragmentDefinitions: true and document has only fragments, text target', () => {
      expect(
        transformAndPrint({
          tsContent: `
            const fragment = hoge\`abc\`;
          `,
          docContent: `
            fragment X on Query {
              helo
            }
          `,
          target: 'text',
          removeFragmentDefinitions: true,
        }),
      ).toMatchSnapshot();
    });

    it('should transform inner document with documentTransformers', () => {
      expect(
        transformAndPrint({
          tsContent: `
          const query = hoge\`abc\`;
        `,
          docContent: `
          query MyQuery {
            hello
          }
          mutation MyMutation {
            bye
          }
        `,
          target: 'object',
          documentTransformers: [
            doc =>
              visit(doc, {
                OperationDefinition: node => (node.operation === 'query' ? node : null),
              }),
          ],
        }),
      ).toMatchSnapshot();
    });
  });

  describe('import declaration transformation', () => {
    describe('default import', () => {
      it('should remove tag import when tag is matched', () => {
        expect(
          transformAndPrint({
            tsContent: 'import hoge from "hoge";',
            tag: 'hoge',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should remove only matched identifier', () => {
        expect(
          transformAndPrint({
            tsContent: 'import hoge, { foo } from "hoge";',
            tag: 'hoge',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore tag import when tag is not matched', () => {
        expect(
          transformAndPrint({
            tsContent: 'import hoge from "hoge";',
            tag: 'foo',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore tag import without tag', () => {
        expect(
          transformAndPrint({
            tsContent: 'import hoge from "hoge";',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });
    });

    describe('named import', () => {
      it('should remove tag import when tag is matched', () => {
        expect(
          transformAndPrint({
            tsContent: 'import { hoge } from "hoge";',
            tag: 'hoge',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should remove alias tag import when tag is matched', () => {
        expect(
          transformAndPrint({
            tsContent: 'import { foo as hoge } from "hoge";',
            tag: 'hoge',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should remove only matched identifier', () => {
        expect(
          transformAndPrint({
            tsContent: 'import foo, { hoge, bar } from "hoge";',
            tag: 'hoge',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore tag import when tag is not matched', () => {
        expect(
          transformAndPrint({
            tsContent: 'import { hoge } from "hoge";',
            tag: 'foo',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore tag import without tag', () => {
        expect(
          transformAndPrint({
            tsContent: 'import { hoge } from "hoge";',
            docContent: '',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should remove * as import tag when tag is matched', () => {
        expect(
          transformAndPrint({
            tsContent: 'import * as hoge from "hoge";',
            docContent: '',
            tag: 'hoge',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore * as import when tag is not matched', () => {
        expect(
          transformAndPrint({
            tsContent: 'import * as hoge from "hoge";',
            docContent: '',
            tag: 'foo',
            target: 'object',
          }),
        ).toMatchSnapshot();
      });
    });
  });

  describe('misc options', () => {
    it('should nothing when getEnabled option returns false', () => {
      expect(
        transformAndPrint({
          tsContent: `const query = \`abc\`;`,
          docContent: ``,
          target: 'object',
          enabled: false,
        }),
      ).toMatchSnapshot();
    });
  });
});
