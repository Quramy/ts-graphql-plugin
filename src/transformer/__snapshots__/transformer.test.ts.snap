// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer GraphQL document transformation should ignore NoSubstitutionTemplateLiteral when tag is set 1`] = `
"const query = \`abc\`;
"
`;

exports[`transformer GraphQL document transformation should ignore TaggedTemplateExpression when the node does not match tag name 1`] = `
"const query = hoge \`abc\`;
"
`;

exports[`transformer GraphQL document transformation should ignore TemplateExpression when tag is set 1`] = `
"const query = \`abc\${def}\`;
"
`;

exports[`transformer GraphQL document transformation should transform NoSubstitutionTemplateLiteral 1`] = `
"const query = {
    kind: \\"Document\\",
    definitions: [{
            kind: \\"OperationDefinition\\",
            operation: \\"query\\",
            variableDefinitions: [],
            directives: [],
            selectionSet: {
                kind: \\"SelectionSet\\",
                selections: [{
                        kind: \\"Field\\",
                        name: {
                            kind: \\"Name\\",
                            value: \\"hello\\"
                        },
                        arguments: [],
                        directives: []
                    }]
            }
        }]
};
"
`;

exports[`transformer GraphQL document transformation should transform TaggedTemplateExpression 1`] = `
"const query = {
    kind: \\"Document\\",
    definitions: [{
            kind: \\"OperationDefinition\\",
            operation: \\"query\\",
            variableDefinitions: [],
            directives: [],
            selectionSet: {
                kind: \\"SelectionSet\\",
                selections: [{
                        kind: \\"Field\\",
                        name: {
                            kind: \\"Name\\",
                            value: \\"hello\\"
                        },
                        arguments: [],
                        directives: []
                    }]
            }
        }]
};
"
`;

exports[`transformer GraphQL document transformation should transform TaggedTemplateExpression when the node matches tag name 1`] = `
"const query = {
    kind: \\"Document\\",
    definitions: [{
            kind: \\"OperationDefinition\\",
            operation: \\"query\\",
            variableDefinitions: [],
            directives: [],
            selectionSet: {
                kind: \\"SelectionSet\\",
                selections: [{
                        kind: \\"Field\\",
                        name: {
                            kind: \\"Name\\",
                            value: \\"hello\\"
                        },
                        arguments: [],
                        directives: []
                    }]
            }
        }]
};
"
`;

exports[`transformer GraphQL document transformation should transform TemplateExpression 1`] = `
"const query = {
    kind: \\"Document\\",
    definitions: [{
            kind: \\"OperationDefinition\\",
            operation: \\"query\\",
            variableDefinitions: [],
            directives: [],
            selectionSet: {
                kind: \\"SelectionSet\\",
                selections: [{
                        kind: \\"Field\\",
                        name: {
                            kind: \\"Name\\",
                            value: \\"hello\\"
                        },
                        arguments: [],
                        directives: []
                    }]
            }
        }]
};
"
`;

exports[`transformer GraphQL document transformation should transform inner document with documentTransformers 1`] = `
"const query = {
    kind: \\"Document\\",
    definitions: [{
            kind: \\"OperationDefinition\\",
            operation: \\"query\\",
            name: {
                kind: \\"Name\\",
                value: \\"MyQuery\\"
            },
            variableDefinitions: [],
            directives: [],
            selectionSet: {
                kind: \\"SelectionSet\\",
                selections: [{
                        kind: \\"Field\\",
                        name: {
                            kind: \\"Name\\",
                            value: \\"hello\\"
                        },
                        arguments: [],
                        directives: []
                    }]
            }
        }]
};
"
`;

exports[`transformer GraphQL document transformation should transform to 0 literal when removeFragmentDefinitions: true and document has only fragments 1`] = `
"const fragment = 0;
"
`;

exports[`transformer GraphQL document transformation should transform to empty string when removeFragmentDefinitions: true and document has only fragments, text target 1`] = `
"const fragment = \\"\\";
"
`;

exports[`transformer GraphQL document transformation should transform to string literal when target is text 1`] = `
"const query = \\"query MyQuery {\\\\n  hello\\\\n}\\\\n\\";
"
`;

exports[`transformer import declaration transformation default import should ignore tag import when tag is not matched 1`] = `
"import hoge from \\"hoge\\";
"
`;

exports[`transformer import declaration transformation default import should ignore tag import without tag 1`] = `
"import hoge from \\"hoge\\";
"
`;

exports[`transformer import declaration transformation default import should remove only matched identifier 1`] = `
"import { foo } from \\"hoge\\";
"
`;

exports[`transformer import declaration transformation default import should remove tag import when tag is matched 1`] = `""`;

exports[`transformer import declaration transformation named import should ignore * as import when tag is not matched 1`] = `
"import * as hoge from \\"hoge\\";
"
`;

exports[`transformer import declaration transformation named import should ignore tag import when tag is not matched 1`] = `
"import { hoge } from \\"hoge\\";
"
`;

exports[`transformer import declaration transformation named import should ignore tag import without tag 1`] = `
"import { hoge } from \\"hoge\\";
"
`;

exports[`transformer import declaration transformation named import should remove * as import tag when tag is matched 1`] = `
"import * as hoge from \\"hoge\\";
"
`;

exports[`transformer import declaration transformation named import should remove alias tag import when tag is matched 1`] = `""`;

exports[`transformer import declaration transformation named import should remove only matched identifier 1`] = `
"import foo, { bar } from \\"hoge\\";
"
`;

exports[`transformer import declaration transformation named import should remove tag import when tag is matched 1`] = `""`;

exports[`transformer misc options should nothing when getEnabled option returns false 1`] = `
"const query = \`abc\`;
"
`;
